File: kotlin-cryptography/src/main/kotlin/me/cryptography/Cryptography.kt
package me.cryptography

import java.util.*
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

class Cryptography {

    fun aes256CbcEncrypt(text: String, password: String): String {
        val saltBytes = randomBytes(16)
        val ivBytes = randomBytes(16)

        val iterationCount = 10000
        val keyLen = 32
        val keyBytes = PBKDF.pbkdf2("HmacSHA256", password.toByteArray(), saltBytes, iterationCount, keyLen)

        val ivSpec = IvParameterSpec(ivBytes)
        val keySpec = SecretKeySpec(keyBytes, "AES")

        val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        cipher.init(Cipher.ENCRYPT_MODE, keySpec, ivSpec);
        val encryptedBytes = cipher.doFinal(text.toByteArray())

        return Base64.getEncoder()
            .encodeToString(saltBytes + ivBytes + encryptedBytes)
            .replace("\n", "")
    }

    fun aes256CbcDecrypt(encrypted: String, password: String): String {
        val concatenatedBytes = Base64.getDecoder().decode(encrypted)

        val saltBytes = concatenatedBytes.copyOfRange(0, 16)
        val ivBytes = concatenatedBytes.copyOfRange(16, 32)
        val encryptedBytes = concatenatedBytes.copyOfRange(32, concatenatedBytes.size)

        val iterationCount = 10000
        val keyLen = 32
        val keyBytes = PBKDF.pbkdf2("HmacSHA256", password.toByteArray(), saltBytes, iterationCount, keyLen)

        val ivSpec = IvParameterSpec(ivBytes)
        val keySpec = SecretKeySpec(keyBytes, "AES")

        val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        cipher.init(Cipher.DECRYPT_MODE, keySpec, ivSpec)
        val decryptedBytes = cipher.doFinal(encryptedBytes)

        return String(decryptedBytes)
    }

    private fun randomBytes(size: Int): ByteArray {
        val random = Random()
        val bytes = ByteArray(size)
        random.nextBytes(bytes)
        return bytes
    }
}

File: nodejs-cryptography/src/cryptography.js
const crypto = require('crypto')

function aes256CbcEncrypt(text, password) {
    const saltBuffer = crypto.randomBytes(16) // 16 bytes
    const ivBuffer = crypto.randomBytes(16) // 16 bytes

    const iterationCount = 10000
    const keyLen = 32 // 256 bits
    const key = crypto.pbkdf2Sync(password, saltBuffer, iterationCount, keyLen, 'sha256')
  
    const cipher = crypto.createCipheriv('aes-256-cbc', key, ivBuffer)
    cipher.write(text)
    cipher.end()
    const encryptedBuffer = cipher.read()
  
    return Buffer.concat([saltBuffer, ivBuffer, encryptedBuffer]).toString('base64')
}

function aes256CbcDecrypt(encrypted, password) {  
    const concatenatedBuffer = Buffer.from(encrypted, 'base64')

    const saltBuffer = concatenatedBuffer.subarray(0, 16)
    const ivBuffer = concatenatedBuffer.subarray(16, 32)
    const encryptedBuffer = concatenatedBuffer.subarray(32)
  
    const iterationCount = 10000
    const keyLen = 32
    const key = crypto.pbkdf2Sync(password, saltBuffer, iterationCount, keyLen, 'sha256')
  
    const decipher = crypto.createDecipheriv('aes-256-cbc', key, ivBuffer)
    decipher.write(encryptedBuffer)
    decipher.end()
    const decrypted = decipher.read()
    return decrypted.toString()
}

module.exports = {
    aes256CbcEncrypt,
    aes256CbcDecrypt
}